astropy/utils/misc.py

```diff
@@ -891,6 +889,27 @@ def set_locale(name):
                 locale.setlocale(locale.LC_ALL, saved)
 
 
+@deprecated('4.0')
+@contextmanager
+def set_locale(name):
+    """Deprecated version of :func:`_set_locale` above.
+    See https://github.com/astropy/astropy/issues/9196
+    """
+    name = str(name)
+
+    with LOCALE_LOCK:
+        saved = locale.setlocale(locale.LC_ALL)
+        if saved == name:
+            # Don't do anything if locale is already the requested locale
+            yield
+        else:
+            try:
+                locale.setlocale(locale.LC_ALL, name)
+                yield
+            finally:
+                locale.setlocale(locale.LC_ALL, saved)
```

Comment:
Is there not a way to avoid duplicating the code? Something like

```python
def set_locale(name):
    yield _set_locale(name)
```

or similar?

Reply:
Nope. I either get a nested contextmanager (not what we want) or `AttributeError: __enter__`. If you find something that works, feel free to push straight here. Otherwise, I don't plan to spending too much time on this.

Reply:
Di dyou try something like this ?
```
set_locale = deprecated('4.0')(_set_locale)
```

Reply:
Oh... that might just work. Thanks, @saimn !